lib/dialects/mysql/connection-manager.js

```diff
@@ -69,82 +85,93 @@ class ConnectionManager extends AbstractConnectionManager {
         }
       }
 
-      const connection = this.lib.createConnection(connectionConfig);
-
-      connection.connect(err => {
-        if (err) {
-          if (err.code) {
-            switch (err.code) {
-            case 'ECONNREFUSED':
-              reject(new sequelizeErrors.ConnectionRefusedError(err));
-              break;
-            case 'ER_ACCESS_DENIED_ERROR':
-              reject(new sequelizeErrors.AccessDeniedError(err));
-              break;
-            case 'ENOTFOUND':
-              reject(new sequelizeErrors.HostNotFoundError(err));
-              break;
-            case 'EHOSTUNREACH':
-              reject(new sequelizeErrors.HostNotReachableError(err));
-              break;
-            case 'EINVAL':
-              reject(new sequelizeErrors.InvalidConnectionError(err));
-              break;
-            default:
-              reject(new sequelizeErrors.ConnectionError(err));
-              break;
-            }
-          } else {
-            reject(new sequelizeErrors.ConnectionError(err));
-          }
-
-          return;
-        }
+      return new Utils.Promise((resolve, reject) => {
+        const connection = this.lib.createConnection(connectionConfig);
+        connection.once('error', (err) => {
+          reject(err);
+        });
+        connection.once('connect', () => {
+          connection.removeAllListeners('error');
```

Comment:
@sushantdhiman the connection promise ought to be resolved at that point though so the reject doesn't matter.

In any case we need to remove both the connect and error listener (atleast the specific ones) on either connect or error so we don't have a leak.